### DISCUSSION TITLE
Describe underlying maths and numerical implementations within vignettes

### DISCUSSION THREAD

**Discussion by TimTaylor** [epiverse-trace/epiverse-trace.github.io#83] [General]
Highlighting this [issue](https://github.com/epiverse-trace/finalsize/issues/176#issue-1813531065) from finalsize as the principle will apply across multiple packages.

Currently the finalsize documentation is missing any real description of the methods employed, both in terms of the underlying mathematics and the numerical solver implementations. Whilst references to papers are given, as a user I would appreciate the package being self contained with all relevant information, including methodology assumptions, available in a vignette.

Do others agree with this requirement? If so - should it be made a "release blocker" (for want of a better term) for CRAN releases of Epiverse packages?

Due to the background required, I suspect these will often require drafting from the PI.

**Comment by jamesmbaazam** [comment]
I'm glad that you asked, Tim. I do see it as a great addition to our methods packages.

I discussed this with Seb sometime back and raised a similar issue on [epichains](https://github.com/epiverse-trace/epichains) (See https://github.com/epiverse-trace/epichains/issues/35). 

I assigned @sbfnk as he originally derived some of the analytical solutions.

**Comment by jamesmbaazam** [comment]
Related to the issue of self-containment, [bpmodels](https://github.com/epiverse-trace/bpmodels) also has [a vignette with a bibliography of applications of branching processes in epidemiology](https://epiverse-trace.github.io/bpmodels/articles/branching_process_literature.html). 

I thought it would be a good idea for users to be able to see how the methods have been applied elsewhere. I have also outlined how users can contribute to the bibliography list. Here is [an example of a PR contributing to the list](https://github.com/epiverse-trace/bpmodels/pull/78).

**Comment by jamesmbaazam** [comment]
> Do others agree with this requirement? If so - should it be made a "release blocker" (for want of a better term) for CRAN releases of Epiverse packages?

I think it's a nice standard to establish.

**Comment by pratikunterwegs** [comment]
> Do others agree with this requirement? If so - should it be made a "release blocker" (for want of a better term) for CRAN releases of Epiverse packages?

I agree that the packages should have their maths and their algorithms documented, but for it to be made a release blocker we'd need to get the PIs or more mathematically minded folks to commit to contributing this documentation.

The methodology assumptions are IMO slightly different, in that those could be written by RSEs without necessarily a full understanding of the maths (see e.g. the "what we assume" callouts in {finalsize} vignettes). These could, by themselves, be very useful to many users. It really depends on who we think the target users are, and whether they will have or want mostly an empirical understanding of our packages ("run these commands in sequence") or a deeper theoretical one as well.

**Comment by TimTaylor** [comment]
> I agree that the packages should have their maths and their algorithms documented, but for it to be made a release blocker we'd need to get the PIs or more mathematically minded folks to commit to contributing this documentation.

Yes - but that is a good thing!

> It really depends on who we think the target users are, and whether they will have or want mostly an empirical understanding of our packages ("run these commands in sequence") or a deeper theoretical one as well.

I think it's best to assume users that want to gain the knowledge and making that as frictionless as possible should be the goal.

**Reply by pratikunterwegs** [reply]
Yes indeed it's a good thing - my point (poorly phrased) was that we should build consensus around this and get the maths folks to agree to doing it before adopting it as a standard, rather than adopting it first and then springing this requirement on them.

**Comment by avallecam** [comment]
> I think it's best to assume users that want to gain the knowledge and making that as frictionless as possible should be the goal.

I agree with this statement. 

I'm thinking about where to add this information, depending on the defined audience for one vignette. 

We can have two types of vignettes:
- "how-to guides"-like vignettes, more focused on steps and code, with the minimally required amount of [concept/method explanations](https://documentation.divio.com/how-to-guides/#don-t-explain-concepts), and
- "explanation"-heavy vignettes, closer to [book/course level details](https://github.com/epiverse-trace/finalsize/issues/176#issuecomment-1643752944) when required.

An additional advantage of this could be in the [review process](https://github.com/epiverse-trace/finalsize/issues/126). It should be easier to track a step/equation in a line of a vignette than in a paper. Alternatively, add numbered equations from papers when available in R files.

**Comment by adamkucharski** [comment]
I think there's a balance to be struck, as a lot of the methodology in our packages will be about numerical efficiency rather than epidemiological concepts. For example, to properly understand what an SIR simulation with {deSolve} is doing, the user would need to know the fundamentals of calculus and numerical integration (e.g. why Runke-Kutta preferable to Euler method). 

To take {finalsize} as an example, the underlying concept is fairly straightforward (i.e. it gives the final size of the epidemic we'd expect from SIR model) but the steps required to get their efficiently require a few tricks (e.g. dividing the SIR ODEs to get dS/dI, integrating the resulting equation, then using a numerical solver like Newton's method). I'd say the {cfr} package is different, because the key delay calculation doesn't require much beyond thinking through how to add up vector elements, so equation more accessible.

When I've taught modelling over the years, I've found that including brief technical details can sometimes be worse than just pointing to further reading for the subset who are interested - it can make the approach more intimidating than it needs to be for occasional users, and can also be unsatisfying (mathematical tricks like the dS/dI step above feel a bit like pulling a rabbit out of a hat). Routinely providing book/course level details would also require a massive time commitment (more than is feasible for many packages).

A few options that may be more doable as routine practice (some of which noted above):
- Give equations briefly where relevant/accessible and point to external resources, e.g. like the [vaccineff pages](https://epiverse-trace.github.io/vaccineff/articles/vaccineff.html), rather than writing a full methodological tutorial. The {bpmodels} [vignette](https://github.com/epiverse-trace/bpmodels/pull/134) also seems like reasonable compromise, giving the equations but not the full generating function theory or derivation. The underlying equations/methods are already coded in the packages, so this would be an easier starting point from a development point of view, then would just need some additional conceptual context and references.
- Include figures to highlight key concepts, like the [EpiEstim pages have](https://mrc-ide.github.io/EpiEstim/articles/full_EpiEstim_vignette.html)
- Feature examples of wider applications and case studies, as well as good reference lists, so users can explore these.
- Point to review papers or introductory concept guides already out there, e.g. [this one from PRISM](http://prism.edu.au/publications/prism-modeling-guideline/)

**Comment by TimTaylor** [comment]
> I think there's a balance to be struck, as a lot of the methodology in our packages will be about numerical efficiency rather than epidemiological concepts. For example, to properly understand what an SIR simulation with {deSolve} is doing, the user would need to know the fundamentals of calculus and numerical integration (e.g. why Runke-Kutta preferable to Euler method).

Agree and the suggestion is not to go in to the exact details of numerical methods ...

> To take {finalsize} as an example, the underlying concept is fairly straightforward (i.e. it gives the final size of the epidemic we'd expect from SIR model) ...

Yes but even the underlying concept nor input assumptions are really described in the {finalsize} documentation

> but the steps required to get their efficiently require a few tricks (e.g. dividing the SIR ODEs to get dS/dI, integrating the resulting equation, then using a numerical solver like Newton's method).

Yes and I really _**do think**_ these details should be in a vignette.

> When I've taught modelling over the years, I've found that including brief technical details can sometimes be worse than just pointing to further reading for the subset who are interested - it can make the approach more intimidating than it needs to be for occasional users, and can also be unsatisfying (mathematical tricks like the dS/dI step above feel a bit like pulling a rabbit out of a hat).

As @avallecam alluded to, R packages can contain multiple vignettes which have different foci (and even different audiences). I often do not read all package vignettes, only the ones I am interested in. My main point is this information should be in a vignette.

>  Routinely providing book/course level details would also require a massive time commitment (more than is feasible for many packages).

If there is remaining funding I think there would be a massive benefit to bring some dedicated modellers in to the group to enable this time commitment.

From my own personal / organisation perspective - EpiVerse packages have a massive opportunity to really stand out and set the standard here. Having a decent level of information contained **_within_** the package documentation rather than hidden behind a linked-list of papers (many paywalled) and books (likely expensive) would be a massive improvement for knowledge accessibility.

**Comment by sbfnk** [comment]
I broadly agree that it's really useful if a package contains the vignette laying out the methodology that's used in the code - also helps users/reviewers checking that the functions do what they're supposed to do.
I recently added a PR with a background vignette to `bpmodels` here in case anyone has comments/suggestions for improvements in this context:
https://github.com/epiverse-trace/bpmodels/pull/134

See also the "Model definitions" vignettes we recently added to EpiNow2:
https://epiforecasts.io/EpiNow2/dev/

**Reply by TimTaylor** [reply]
Yes - they're exactly the sort of thing that I was thinking for content (minor pr incoming for bpmodels theory)

**Comment by adamkucharski** [comment]
Yep, like the bpmodels example. Agree good to improve access to open access materials on methods where possible. As our training materials expand – which will have dedicated resources – alongside links with organisations focused on open training (like AppliedEpi), it should become clearer what sits best where. Expect there will also be quite a lot of overlap in places, so can decide whether replication or linking makes more sense.

**Comment by adamkucharski** [comment]
Started a draft with finalsize theory vignette (fortunately had the equations latexed up in some old draft documents – may not be so lucky with other packages!) Will add varying susceptibility shortly. Hard to avoid lots of integrals (especially as don't think we can use the dS/dR trick for multiple age groups) https://github.com/epiverse-trace/finalsize/commit/88808cd0c6971a4ee58d06880b3a430483fd9a93

**Reply by pratikunterwegs** [reply]
Thanks, if you could ping me when complete I can link to this from the other vignettes too

**Reply by Bisaloo** [reply]
As an aside, [Mathpix](https://mathpix.com/) is quite good to convert equations from papers or handwritten equations to LaTeX / Mathjax.

**Reply by adamkucharski** [reply]
I've updated the vignette – might be good to get a fresh pair of eyes on it from @BlackEdder before doing a PR, just to make sure the version of heterogeneous susceptibility I've outlined is appropriated given the code implementation.

**Comment by joshwlambert** [comment]
During the UK Epidemic Analytics and Modelling Event on February 13th & 14th 2025 we hosted an Epiverse-TRACE session, within which we collected feedback from participants on Epiverse tools. One comment pertinent to this thread was:

> ... useful to have the theory easily accessible alongside vignettes (like in finalsize)

Therefore, it looks like the discussion earlier in this thread produced some math and implementation documentation that others in the UK epidemic modelling field appreciate. It would be good to identify other Epiverse-TRACE packages that could benefit from adding similar documentation.
